[flang][debug] Generate DISubprogramAttr for omp::TargetOp.

abidh opened the pull request:
There are `DeclareOp` present for the variables mapped into target region. That allow us to generate debug information for them. Bu the `TargetOp` is still part of parent function and those variables get the parent function's `DISubprogram` as a scope.

In `OMPIRBuilder`, a new function is created for the `TargetOp`. We also create a new `DISubprogram` for it. All the variables that were in the target region now have to be updated to have the correct scope. This after the fact updating of debug information becomes very difficult in certain cases. Take the example of variable arrays. The type of those arrays depend on the artificial `DILocalVariable`(s) which hold the size(s) of the array. This new function will now require that we generate the new variable and and new types. Similar issue exist for character type variables too.

To avoid this after the fact updating, this PR generates a `DISubprogramAttr` for the `TargetOp` while generating the debug info in flang. This help us avoid updating later.

This PR is flang side of the change. The https://github.com/llvm/llvm-project/pull/138149 makes the required changes in `OMPIRBuilder`.

abidh committed:
[flang][debug] Generate DISubprogramAttr for omp::TargetOp.

There are DeclareOp present for the variables mapped into target region. That
allow us to generate debug information for them. Bu the TargetOp is still
part of parent function and those variables get the parent function's
DISubprogram as a scope.

In OMPIRBuilder, a new function is created for the TargetOp. We also create a
new DISubprogram for it. All the variables that were in the target region now
have to be updated to have the correct scope. This after the fact updating of
debug information becomes very difficult in certain cases. Take the example of
variable arrays. The type of those arrays depend on the artificial
DILocalVariable(s) which hold the size(s) of the array. This new function will
now require that we generate the new variable and and new types. Similar
issue exist for character type variables too.

To avoid this after the fact updating, this PR generates a DISubprogramAttr
for the TargetOp while generating the debug info in flang. This help us
avoid updating later.

This PR is flang side of the change. I will open another PR which will make
the required changes in OMPIRBuilder.

abidh commented:
Polite ping.

skatrak commented:
Thank you Abid, can you check buildbot failures? They seem to be related to the patch.

abidh commented:
> Thank you Abid, can you check buildbot failures? They seem to be related to the patch.
Hi Sergio,
Thanks for having a look. The failures are expected as this PR needs the accompanying https://github.com/llvm/llvm-project/pull/138149 for build/tests to work correctly. I kept them separate for ease of reviewing. Once both of them are approved, I will bring them together before merge to make sure that build bots are green and they go in as one commit.

skatrak commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -510,8 +542,73 @@ void AddDebugInfoPass::handleFuncOp(mlir::func::FuncOp funcOp,
+      }
+    }
+
+    Id = mlir::DistinctAttr::create(mlir::UnitAttr::get(context));
```
I wonder whether it's actually necessary to create new attributes to hold the same constant value. I might be wrong on that, just thinking that MLIR attributes are already uniqued at the context level, so my impression is that this would just end up pointing to the same thing in the end. But I may have misunderstood that, so not a request to change anything.

abidh replied:
I will do further investigation on it because the similar code is used in other places too.

skatrak commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -103,6 +104,37 @@ bool debugInfoIsAlreadySet(mlir::Location loc) {
+// getTargetEntryUniqueInfo and
+// TargetRegionEntryInfo::getTargetRegionEntryFnName to generate the same name.
+// But even if there was a slight mismatch, it is not a problem because this
+// name is artifical and not important to debug experience.
```
```suggestion
// name is artificial and not important to debug experience.
```

skatrak reviewed (commented):
This seems fine to me, I just have a couple of nits and potentially basic questions. Thanks again!

abidh committed:
Handle review comments.

Main change is that we also take care of the case when only line table
information is requested.

abidh committed:
Handle review comments.

Use the arguments to the TargetOp in the subroutine type created for it.

abidh commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
That is a good catch. I have re-arranges the code so that we handle both cases(lineTableOnly and FullDebug). I added a testcase for it too.

abidh committed:
Fix formatting issues.

abidh commented:
Polite ping.

skatrak reviewed (commented):
Thank you Abid, and excuse me for the delay giving your changes another look. This mostly LGTM, I just have one remaining concern.

abidh committed:
Handle review comments.

skatrak commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -510,8 +542,73 @@ void AddDebugInfoPass::handleFuncOp(mlir::func::FuncOp funcOp,
+    mlir::DistinctAttr Id =
+        mlir::DistinctAttr::create(mlir::UnitAttr::get(context));
+    llvm::SmallVector<mlir::LLVM::DITypeAttr> types;
+    types.push_back(mlir::LLVM::DINullTypeAttr::get(context));
```
In that case, I'd suggest just passing a single `DINullTypeAttr` (as if it was a parameter-less void function) and then adding a comment explaining what you're saying, basically stating that figuring out the actual argument list of the outlined function at that point is non-trivial and we don't actually need the function type information to be correct to produce the right dwarf. I think this was your original approach, minus the comment, so sorry about the back and forth before agreeing with you! 😄

That way we don't do any work to fill out the argument types incorrectly anyways, and we keep this decision documented in case we later need to actually fill the argument types properly.

skatrak reviewed (approved):
LGTM, thank you!
